### PR TITLE
cgen, check, parser: implement first-class option type

### LINKED ISSUE #16861: cgen: cannot convert 'struct string' to 'struct _option_string'
### Describe the bug

casting problem

### Expected Behavior

no error

### Current Behavior

error

### Reproduction Steps

```v
struct StructType {
mut:
	a string
	b ?string
}

struct Decoder {}

fn (d &Decoder) decode[T]() T {
	typ := T{}
    $if T is $Struct {
		dump("is Struct")
        $for field in T.fields {
			$if field.is_optional {
				typ.$(field.name) = "a"
			} $else {
				typ.$(field.name) = "b"
			}
        }
    }
	return typ
}

fn main() {
	d := Decoder{}
    result := d.decode[StructType]()
	println(result)
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.2 0a6fc6d

### Environment details (OS name and version, etc.)

Ubuntu

### LINKED ISSUE #16862: checker: error: cannot assign a `none` value to a variable
### Describe the bug

cannot assign a `none` value to a struct field

### Expected Behavior

```v
[b.v:12] 'is Struct': is Struct
StructType{
    a: 'b'
    b: none
}
```

### Current Behavior

```bash
b.v:15:25: error: cannot assign a `none` value to a variable
   13 |         $for field in T.fields {
   14 |             $if field.is_optional {
   15 |                 typ.$(field.name) = none
      |                                     ~~~~
   16 |             } $else {
   17 |                 typ.$(field.name) = "b"
```

### Reproduction Steps

```v
struct StructType {
mut:
	a string
	b ?string
}

struct Decoder {}

fn (d &Decoder) decode[T]() T {
	typ := T{}
    $if T is $Struct {
		dump("is Struct")
        $for field in T.fields {
			$if field.is_optional {
				typ.$(field.name) = none
			} $else {
				typ.$(field.name) = "b"
			}
        }
    }
	return typ
}

fn main() {
	d := Decoder{}
    result := d.decode[StructType]()
	println(result)
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.2 0a6fc6d

### Environment details (OS name and version, etc.)

Ubuntu

### PR DESCRIPTION
**Pull request by felipensp**
Fix #16862
Fix #16861

Makes possible to do:

## Option propagation
```V
fn test(a ?int) ?int {
	var := a? // = or { return none }
	println('this will not be printed')
	assert false
	println(var)
	return 123 // dont get here
}

fn main() {
	a := ?int(none)
	b := test(a) // none
	println(b) // Option(error: none)
}
```

## Option return propagation 
```V
fn test() ?int { return none }
fn main() {
  var := test()
  assert var == none
}
```

## Option initialization 
```V
mut x = ?int(123)
mut y = x
y = none
assert y == none
```

## Option unwrapping
```V
var := ?int(1)
println(var?) // i.e. var or { return none }
assert var? == 1

var2 := var? + 1
assert var2 == 2
println('${var2}') // 2
```

### Option type
```V
struct Foo {
mut:
	name string
}

fn (mut f Foo) test(arr ?[]int) ?string {
	return arr?.len.str()
}

fn main() {
	mut m := ?Foo( Foo{} ) // just `mut m := ?Foo{}` means that `m` will be initialized to `none`
	assert m?.test([1,2,3])? == '3'
	m?.name = 'foo'
	assert m?.name == 'foo'
}
```

### Option assertion
```V
mut x := ?int(none)
assert x == none
x = 123
assert x != none
assert x? == 123
```

### Unset option var
```V
typ.$(field.name) = none
```

### Option as function argument
```V
fn is_none(a ?string) bool {
	return a == none
}

fn is_map_none(a ?map[string]int) bool {
	return a == none
}

fn test_opt_call_arg() {
	assert is_none(?string(none))
	var := ?string('foo')
	assert is_none(var) == false

	mut var2 := ?map[string]int{}
	assert var2 == none
	var2?['a'] = 1
	assert is_map_none(var2) == false
}
```

```V
struct Call {
pub:
	a int
}

fn testcall(p ?Call) ?Call {
	return p or { Call{a:1}}
}

fn main() {
	println(testcall(none)) // Option(Call{a: 1})
}
```

### Default type mismatch checking
```
vlib/v/checker/tests/var_option_wrong_default.vv:4:19: error: `or` block must provide a value of type `int`, not `?int`
    2 | 
    3 | var2 := var1 or { var1? }
    4 | var3 := var1 or { var1 }
      |                   ~~~~
```
```
vlib/v/checker/tests/var_option_wrong_type.vv:3:23: error: mismatched types `?f64` and `string`
    1 | var_none := ?f64(none)
    2 | 
    3 | var3 := var_none or { var_none + 'foo' }
      |                       ~~~~~~~~~~~~~~~~
```

### CODE REVIEW

**Comment by spytheman** [comment]
Is there something missing for it to be ready to review?

File: vlib/v/checker/checker.v [hunk 1]
```diff
@@ -3834,8 +3871,13 @@ fn (mut c Checker) index_expr(mut node ast.IndexExpr) ast.Type {
 		c.error('type `${typ_sym.name}` does not support indexing', node.pos)
 	}
 	if typ.has_flag(.option) {
-		c.error('type `?${typ_sym.name}` is an option, it does not support indexing',
-			node.left.pos())
+		if node.left is ast.Ident && (node.left as ast.Ident).or_expr.kind == .absent {
+			c.error('type `?${typ_sym.name}` is an option, it must be unwrapped `var?[]`',
```
Comment (spytheman) [inline vlib/v/checker/checker.v:3875]:
```suggestion
			c.error('type `?${typ_sym.name}` is an Option, it must be unwrapped first; use `var?[]` to do it',
```

File: vlib/v/checker/checker.v [hunk 2]
```diff
@@ -3834,8 +3871,13 @@ fn (mut c Checker) index_expr(mut node ast.IndexExpr) ast.Type {
...
+				node.left.pos())
+		} else if node.left is ast.CallExpr {
+			c.error('type `?${typ_sym.name}` is an option, it must be unwrapped `func()` or use or-block `or {}`',
```
Comment (spytheman) [inline vlib/v/checker/checker.v:3878]:
```suggestion
			c.error('type `?${typ_sym.name}` is an option, it must be unwrapped with `func()?`, or use `func() or {default}`',
```

File: vlib/v/tests/option_var_map_test.v [hunk 3]
```diff
@@ -0,0 +1,14 @@
+fn abc(a ?map[string]string) ?string {
+	return a?['foo']
+}
+
+fn abc_2(a ?map[string]string) string {
+	return a?['foo']
```
Comment (spytheman) [inline vlib/v/tests/option_var_map_test.v:6]: This should not be allowed, since the function does not return an option type, it returns `string`. The reason is that when `a` is `none`, `a?` will try to propagate that to the caller, but the caller does not expect `none` at all (it expects to get a string).

Instead, this should work: `return (a or { map[string]string{} })['foo']`, and also this:
```v
if b := a {
    return b['foo']
}
```

File: vlib/v/checker/checker.v (hunk 2 above)
Comment (JalonSolov) [reply vlib/v/checker/checker.v:3878]: `Option` (instead of `option`) here, too.

File: vlib/v/tests/option_var_map_test.v [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+fn test_option_map() {
+	mut var4 := ?map[string]string{}
+	var4?['foo'] = 'foo'
```
Comment (spytheman) [inline vlib/v/tests/option_var_map_test.v:11]: Here that should not work too, since `var4` was `none` (the default value for all option types).

Note that `var4?` is allowed here, since `test_` functions are exception to the normal V rule of disallowing using `!` and `?` to propagate errors and nones (in test_ functions, the propagation just acts as a failed assertion).

File: vlib/v/tests/option_var_test.v [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+	})
+	println(result)
+}
+
+fn test_cast_option() {
+	mut x := ?int(123)
+	dump(x)
+	assert x != none
+	x = none
+	dump(x)
```
Comment (spytheman) [inline vlib/v/tests/option_var_test.v:47]: an additional `assert x == none` here will be good too

File: vlib/v/tests/option_var_test.v [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+	assert y != none
+}
+
+fn test_blank_assign() {
+	_ := ?bool(false)
+}
+
+fn test_optional_value_assign() {
+	x := ?int(0)
+	assert x != none
```
Comment (spytheman) [inline vlib/v/tests/option_var_test.v:64]: add also `assert x? == 0`

File: vlib/v/tests/option_var_test.v [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+	x = none
+	dump(x)
+}
+
+fn test_assign_from_option() {
+	mut x := ?int(123)
+	mut y := x
+	println(y)
+	assert x != none
+	assert y != none
```
Comment (spytheman) [inline vlib/v/tests/option_var_test.v:55]: add also `assert y? == 123`

File: vlib/v/tests/option_var_test.v [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+	assert var3 == none
+
+	var = 1
+	var2 = 2
+	var3 = 3
+
+	assert var != none
+	assert var2 != none
+	assert var3 != none
+}
```
Comment (spytheman) [inline vlib/v/tests/option_var_test.v:106]: add positive assertions too like `assert var? == 1` etc

File: vlib/v/tests/option_var_test.v [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+fn test_as_cast() {
+	var := StructType2{}
+	b := var.a as ?time.Time
+}
+
+fn test_unwrap() {
+	var := ?int(1)
+	println(var)
+	println(var)
```
Comment (spytheman) [inline vlib/v/tests/option_var_test.v:116]: The second println looks superfluous; perhaps you meant `println(var?)` ?

File: vlib/v/tests/option_var_test.v [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+	var2 := var1 or { 0 }
+	var3 := var1 or { 1 }
+	assert var2 + var3 == 1
+	var4 := var1 or {
+		t := 1 + var3
+		t
+	}
+
+	assert var4 == 2
+}
```
Comment (spytheman) [inline vlib/v/tests/option_var_test.v:136]: Excellent!

File: vlib/v/tests/option_var_test.v [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+fn test_default_values() {
+	var_none := ?f64(none)
+	var_none2 := ?string(none)
+
+	var2 := var_none or { var_none? + 1 }
```
Comment (spytheman) [inline vlib/v/tests/option_var_test.v:142]: I do not get it - `var_none` is `none`, how is it used in the or block ?
`var_none?` should make the test function to abort here.

File: vlib/v/tests/option_var_test.v [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+	println(var2)
+	assert var2 == 1.0
+
+	var3 := var_none2 or { var_none2? + 'foo' }
```
Comment (spytheman) [inline vlib/v/tests/option_var_test.v:146]: Same - `var_none2?` here should abort the test function and cause the entire test file to fail.

File: vlib/v/tests/option_var_test.v [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+	println(var3)
+	assert var3 == 'foo'
+}
+
+fn test_assert_option() {
+	var1 := ?int(none)
+	varz := ?f64(none)
+	assert var1 == none
+
+	var2 := var1 or { var1? }
```
Comment (spytheman) [inline vlib/v/tests/option_var_test.v:156]: same - `?` means propagate the none to the caller, and in the case of test_ functions , that means abort/failure.

File: vlib/v/tests/var_option_arr_test.v [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+fn test_main() {
+	mut var2 := abc()?
+	assert var2.len == 3
+
+	var := ?[]int{len: 10, init: 2}
+	assert var?.len == 10
+	assert var != none
+
+	var3 := ?[]string{}
+	assert var3 == none
```
Comment (spytheman) [inline vlib/v/tests/var_option_arr_test.v:24]: i.e. not using any special initialization flags, means that you use the default `none`, but adding them, means that you initialize the array instead and wrap it in an option ?

File: vlib/v/tests/var_option_struct_test.v [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+	return arr?.len.str()
+}
+
+fn (mut f Foo) test2() ?string {
+	return none
+}
+
+fn test_main() {
+	mut m := ?Foo{}
+	v := m?.test([1, 2, 3]) or { '4' }
```
Comment (spytheman) [inline vlib/v/tests/var_option_struct_test.v:16]: That should be a runtime error - `m` was initialized to `none`, and then `m?` should have propagated that and made the test to fail.

Instead: ```v
mut m := ?Foo{}
m = Foo{}
```
should be ok, so that `m?` later just unwraps and gets the proper `Foo{}` value.

File: vlib/v/tests/var_option_struct_test.v [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+	m?.name = 'foo'
```
Comment (spytheman) [inline vlib/v/tests/var_option_struct_test.v:17]: same - `m` is still `none` , so `m?` here should have caused a runtime error.

File: vlib/v/tests/option_var_test.v (hunk 11 above)
Comment (felipensp) [reply vlib/v/tests/option_var_test.v:142]: Ok, fixed.

```V
bug.v:4:20: error: `or` block must provide a value of type `f64`, not possible none value
    2 |     varz := ?f64(none)
    3 |     assert varz == none
    4 |     var2 := varz or { varz? }
      |                       ~~~~
    5 | }
 ```

File: vlib/v/tests/option_var_test.v (hunk 12 above)
Comment (felipensp) [reply vlib/v/tests/option_var_test.v:146]: Fixed.

File: vlib/v/tests/option_var_test.v (hunk 13 above)
Comment (felipensp) [reply vlib/v/tests/option_var_test.v:156]: Fixed.

File: vlib/v/tests/var_option_arr_test.v (hunk 14 above)
Comment (felipensp) [reply vlib/v/tests/var_option_arr_test.v:24]: since it is prefixed with '?' it is an option array string, even with special init flags. should we change it?

File: vlib/v/checker/checker.v [hunk 17]
```diff
@@ -3877,10 +3877,10 @@ fn (mut c Checker) index_expr(mut node ast.IndexExpr) ast.Type {
 	}
 	if typ.has_flag(.option) {
 		if node.left is ast.Ident && (node.left as ast.Ident).or_expr.kind == .absent {
-			c.error('type `?${typ_sym.name}` is an Option, it must be unwrapped first; use `var?[]` to do it',
+			c.error('type `?${typ_sym.name}` is an option, it must be unwrapped first; use `var?[]` to do it',
```
Comment (JalonSolov) [inline vlib/v/checker/checker.v:3876]: Why the change to lowercase?  It is referring to the official Option time, not a general option.

Comment (felipensp) [reply vlib/v/checker/checker.v:3876]: Sorry, I mistake the point.

File: vlib/v/tests/option_var_test.v (hunk 12 above)
Comment (felipensp) [reply vlib/v/tests/option_var_test.v]: Fixed

File: vlib/v/tests/option_var_test.v (hunk 11 above)
Comment (felipensp) [reply vlib/v/tests/option_var_test.v]: Fixed

File: vlib/v/tests/var_option_struct_test.v (hunk 16 above)
Comment (felipensp) [reply vlib/v/tests/var_option_struct_test.v:38]: Fixed, thanks.

File: vlib/v/tests/var_option_struct_test.v (hunk 15 above)
Comment (felipensp) [reply vlib/v/tests/var_option_struct_test.v:37]: Fixed, thanks.

File: vlib/v/tests/var_option_arr_test.v (hunk 14 above)
Comment (felipensp) [reply vlib/v/tests/var_option_arr_test.v]: Fixed, thanks.

**Comment by spytheman** [comment]
Is there something that remains to be done, for this draft to be ready for review (beside rebasing over current master/solving the conflict) ?

**Comment by felipensp** [comment]
> Is there something that remains to be done, for this draft to be ready for review (beside rebasing over current master/solving the conflict) ?

I've updated the PR. It must be reviewed to check if it is missing some detail to cover about option.

**Comment by spytheman** [comment]
I will review it asap.

File: vlib/v/gen/c/assign.v [hunk 18]
```diff
@@ -7,6 +7,85 @@ import v.ast
...
+			g.expr(expr)
+		}
+		g.writeln(' }, (_option*)(&${tmp_var}), sizeof(${styp}));')
+		g.write(stmt_str)
+		g.write(tmp_var)
+		return tmp_var
+	}
+}
+
+// expr_with_opt is used in assign expr to `optinal`.
```
Comment (spytheman) [inline vlib/v/gen/c/assign.v:65]:
```suggestion
// expr_with_opt is used in assigning an expression to an `option` variable
```

File: vlib/v/gen/c/assign.v [hunk 19]
```diff
@@ -7,6 +7,85 @@ import v.ast
...
+// e.g. x = y (option lhs and rhs), mut x = ?int(123), y = none
+fn (mut g Gen) expr_with_opt(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type) string {
+	if expr_typ == ast.none_type {
+		g.inside_opt_or_res = true
```
Comment (spytheman) [inline vlib/v/gen/c/assign.v:69]: in all other places, where `g.inside_opt_or_res` is set, the previous value is stored, and then it is restored; imho it should be done here as well, otherwise nested expressions will be a problem

File: vlib/v/gen/c/assign.v [hunk 20]
```diff
@@ -7,6 +7,85 @@ import v.ast
...
+	}
+	if expr_typ.has_flag(.option) && ret_typ.has_flag(.option)&& (expr in [ast.Ident, ast.ComptimeSelector, ast.AsCast, ast.CallExpr, ast.MatchExpr, ast.IfExpr, ast.IndexExpr, ast.UnsafeExpr, ast.CastExpr]) {
+		if expr in [ast.Ident, ast.CastExpr] {
+			if expr_typ.idx() != ret_typ.idx() {
+				return g.expr_opt_with_cast(expr, expr_typ, ret_typ)
+			}
+			g.inside_opt_or_res = true
```
Comment (spytheman) [inline vlib/v/gen/c/assign.v:76]: same, please add a defer {} block for restoring `g.inside_opt_or_res`

File: vlib/v/gen/c/assign.v [hunk 21]
```diff
@@ -7,6 +7,85 @@ import v.ast
...
+		}
+		g.expr(expr)
+		return expr.str()
+	} else {
+		g.inside_opt_or_res = true
```
Comment (spytheman) [inline vlib/v/gen/c/assign.v:81]: same, add a `defer { g.inside_opt_or_res = old_inside_opt_or_res }` after this

File: vlib/v/gen/c/assign.v [hunk 22]
```diff
@@ -441,28 +536,11 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
...
-					}
-					return
-				}
-				*/
-			}
 			if !cloned {
 				if !g.inside_comptime_for_field
 					&& ((var_type.has_flag(.option) && !val_type.has_flag(.option))
 					|| (var_type.has_flag(.result) && !val_type.has_flag(.result))) {
+					g.inside_opt_or_res = true
```
Comment (spytheman) [inline vlib/v/gen/c/assign.v:543]: same

File: vlib/v/tests/option_var_test.v [hunk 23]
```diff
@@ -0,0 +1,210 @@
+import time
+
+type MyAlias = []int
+type MySumType = []f64 | []int
+
+struct Test {
+}
+
+struct StructType2 {
+	a ?time.Time = none
```
Comment (spytheman) [inline vlib/v/tests/option_var_test.v:10]: this will produce a warning, `= none` is not needed.

File: vlib/v/gen/c/assign.v (hunk 21 above)
Comment (felipensp) [reply vlib/v/gen/c/assign.v:90]: Done.

File: vlib/v/gen/c/assign.v (hunk 20 above)
Comment (felipensp) [reply vlib/v/gen/c/assign.v:81]: Done.

File: vlib/v/gen/c/assign.v (hunk 19 above)
Comment (felipensp) [reply vlib/v/gen/c/assign.v:70]: Done.

File: vlib/v/gen/c/assign.v (hunk 22 above)
Comment (felipensp) [reply vlib/v/gen/c/assign.v:555]: Done.

File: vlib/v/tests/option_var_test.v (hunk 23 above)
Comment (felipensp) [reply vlib/v/tests/option_var_test.v]: Done.

**Review by spytheman — APPROVED** [review]
